src/mlpack/core/optimizers/cmaes/cmaes_impl.hpp

```diff
@@ -0,0 +1,260 @@
+  const double cs = (muEffective + 2.0) / (iterate.n_elem + muEffective + 5.0);
+  const double ds = 1 + cs + 2 * std::max(std::sqrt((muEffective - 1) /
+      (iterate.n_elem + 1)) - 1, 0.0);
+  const double enn = std::sqrt(iterate.n_elem) * (1.0 - 1.0 /
+      (4.0 * iterate.n_elem) + 1.0 / (21 * std::pow(iterate.n_elem, 2)));
+
+  // Covariance update parameters.
+  // Cumulation for distribution.
+  const double cc = (4.0 + muEffective / iterate.n_elem) /
+      (4 + iterate.n_elem + 2 * muEffective / iterate.n_elem);
+  const double h = (1.4 + 2.0 / (iterate.n_elem + 1.0)) * enn;
+
+
+  const double c1 = 2 / (std::pow(iterate.n_elem + 1.3, 2) + muEffective);
+  const double alphaMu = 2;
+  const double cmu = std::min(1 - c1, alphaMu * (muEffective - 2 + 1 /
+      muEffective) / (std::pow(iterate.n_elem + 2, 2) +
+      alphaMu * muEffective / 2));
+
+  arma::cube mPosition(iterate.n_rows, iterate.n_cols, 3);
+  mPosition.slice(0) = lowerBound + arma::randu(
+      iterate.n_rows, iterate.n_cols) * (upperBound - lowerBound);
+
+  arma::mat step = arma::zeros(iterate.n_rows, iterate.n_cols);
+
+  double currentObjective = function.Evaluate(mPosition.slice(0), 0);
+  for (size_t f = 1; f < numFunctions; ++f)
+    currentObjective += function.Evaluate(mPosition.slice(0), f);
+
+  double overallObjective = currentObjective;
+  double lastObjective = DBL_MAX;
+
+  // Population parameters.
+  arma::cube pStep(iterate.n_rows, iterate.n_cols, lambda);
+  arma::cube pPosition(iterate.n_rows, iterate.n_cols, lambda);
+  arma::vec pObjective(lambda);
+  arma::cube ps = arma::zeros(iterate.n_rows, iterate.n_cols, 3);
+  arma::cube pc = arma::zeros(iterate.n_rows, iterate.n_cols, 3);
+  arma::cube C(iterate.n_elem, iterate.n_elem, 3);
+  C.slice(0).eye();
+
+  // Covariance matrix parameters.
+  arma::vec eigval;
+  arma::mat eigvec;
+  arma::vec eigvalZero = arma::zeros(iterate.n_elem);
+
+  // The current visitation order (sorted by population objectives).
+  arma::uvec idx = arma::linspace<arma::uvec>(0, lambda - 1, lambda);
+
+  // Now iterate!
+  for (size_t i = 1; i < maxIterations; ++i)
+  {
+    // To keep track of where we are.
+    const size_t idx0 = (i - 1) % 3;
+    const size_t idx1 = i % 3;
+
+    const arma::mat covLower = arma::chol(C.slice(idx0), "lower");
+
+    for (size_t j = 0; j < lambda; ++j)
+    {
+      if (iterate.n_rows > iterate.n_cols)
+      {
+        pStep.slice(idx(j)) = covLower *
+            arma::randn(iterate.n_rows, iterate.n_cols);
+      }
+      else
+      {
+        pStep.slice(idx(j)) = arma::randn(iterate.n_rows, iterate.n_cols) *
+            covLower;
+      }
+
+      pPosition.slice(idx(j)) = mPosition.slice(idx0) + sigma(idx0) *
+          pStep.slice(idx(j));
+
+      pObjective(idx(j)) = function.Evaluate(pPosition.slice(idx(j)), 0);
+      for (size_t f = 1; f < numFunctions; ++f)
+        pObjective(idx(j)) += function.Evaluate(pPosition.slice(idx(j)), f);
```

Comment:
I guess when we have batch support we can change this to a single call to `Evaluate()` with batch size `numFunctions`.

Reply:
Agreed, not only here but in the CNE code as well.